File: iosApp/iosApp/features/workout/history/mocks/WorkoutHistoryDataSourceMocks.swift
//
//  WorkoutHistoryDataSourceMocks.swift
//  ShapeByte
//
//  Created by Lang, Stefan [Shape Byte Tech] on 02.08.24.
//

import Foundation
import Combine
import shared
// TODO: move to shared
class WorkoutHistoryDataSourceMocks {
    func historyFor(date: Date, untilPastDate pastDate: Date) -> [WorkoutScheduleEntry] {
        var history: [WorkoutScheduleEntry] = .init()

        let calendar = Calendar.current // Aktueller Kalender
        let components = calendar.dateComponents([.day], from: pastDate, to: date)
        var daysCount = components.day ?? 0

        for currDayIndex in 0..<daysCount {
            guard let date = calendar.date(byAdding: .day, value: (currDayIndex * -1), to: date) else {
                continue
            }

            let entry = WorkoutScheduleEntry(
                id: "\(currDayIndex)",
                name: "Workout \(currDayIndex + 1)",
                date: date,
                progress: Progress(progress: 1)
            )
            history.append(entry)
        }

        return history
    }
}

File: shared/src/commonMain/kotlin/de/stefan/lang/shapebyte/features/workout/history/data/mocks/WorkoutHistoryDataSourceMocks.kt
package de.stefan.lang.shapebyte.features.workout.history.data.mocks

import de.stefan.lang.shapebyte.features.workout.history.data.WorkoutHistoryDataSource
import de.stefan.lang.shapebyte.features.workout.schedule.data.WorkoutScheduleEntry
import de.stefan.lang.shapebyte.shared.loading.data.LoadState
import de.stefan.lang.shapebyte.utils.Progress
import kotlinx.datetime.Instant
import kotlin.time.Duration.Companion.days

object WorkoutHistoryDataSourceMocks : WorkoutHistoryDataSource {
    override suspend fun historyForDates(
        date: Instant,
        pastDate: Instant,
    ): LoadState.Result<List<WorkoutScheduleEntry>> {
        val history = createHistoryEntries(date, pastDate)
        return LoadState.Success(history)
    }

    fun createHistoryEntries(date: Instant, pastDate: Instant): List<WorkoutScheduleEntry> {
        val history = mutableListOf<WorkoutScheduleEntry>()
        val daysCount = date.minus(pastDate).inWholeDays

        for (currDayIndex in 0 until daysCount) {
            val entry = WorkoutScheduleEntry(
                id = "$currDayIndex",
                name = "Workout ${currDayIndex + 1}",
                date = date.minus(currDayIndex.days),
                progress = Progress.COMPLETE,
            )
            history.add(entry)
        }

        return history
    }
}
